### PR TITLE
deprecate_settings: Allow multiple deprecations

### PR DESCRIPTION
**Pull request by pratyushprakash**
Add support for multiple deprecations of the
same setting.

Closes https://github.com/coala/coala/issues/3537

### CODE REVIEW

**Comment by namanyadav12** [comment]
I think you should add tests to show that your changes work.

**Comment by pratyushprakash** [comment]
I have added tests.

**Comment by namanyadav12** [comment]
Oh. Didn't know that.
Thanks for the info.

File: coalib/bearlib/__init__.py [hunk 1]
```diff
@@ -82,34 +96,49 @@ def _deprecate_decorator(func):
...
+                if not isinstance(depr_args_and_modifier, tuple):
+                    depr_args_and_modifier = (depr_args_and_modifier,)
+
+                if callable(depr_args_and_modifier[-1]):
+                    deprecated_args = depr_args_and_modifier[:-1]
+                    _func = depr_args_and_modifier[-1]
+                else:
+                    deprecated_args = depr_args_and_modifier
+
+                    def _func(x): return x
```
Comment (satwikkansal) [inline coalib/bearlib/__init__.py]: I'm wondering what this line does?

Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: We need a function to modify the deprecated arg.

Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: @satwikkansal `depr_arg_value = _func.__call__(kwargs[deprecated_arg])`

File: coalib/bearlib/__init__.py [hunk 2]
```diff
@@ -64,12 +64,26 @@ def deprecate_settings(**depr_args):
     >>> run(old='Hello!', new='Hello!')
     Hello!
 
+    Multiple deprecations can be provided for the same setting. Note that
+    the modifier function if required must always be given as the last element
+    of the deprecated settings tuple.
+
+    >>> @deprecate_settings(new=('old','older'))
```
Comment (sils) [inline coalib/bearlib/__init__.py]: uhm what if I have two different conversion functions?

**Comment by Asnelchristian** [comment]
this looks good to me

File: coalib/bearlib/__init__.py (hunk 2 above)
Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: How would you like to handle this?

Comment (sils) [inline coalib/bearlib/__init__.py]: maybe alllow a dict so for every old value you can give a conversion function or just None for the `lambda x:x`

File: coalib/bearlib/__init__.py [hunk 3]
```diff
@@ -64,12 +64,27 @@ def deprecate_settings(**depr_args):
     >>> run(old='Hello!', new='Hello!')
     Hello!
 
+    Multiple deprecations can be provided for the same setting. The modifier
+    function for each deprecated setting can be given as a value in  a dict
+    where the deprecated setting is the key. A default modifier may be
+    specified at the end of the deprecated settings tuple.
+
+    >>> @deprecate_settings(new=({'old': lambda x: x + ' coala!'},'older'))
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: one space needed before `'older'`

File: coalib/bearlib/__init__.py [hunk 4]
```diff
@@ -82,34 +97,57 @@ def _deprecate_decorator(func):
...
+                if not isinstance(depr_args_and_modifier, tuple):
+                    depr_args_and_modifier = (depr_args_and_modifier,)
+
+                if callable(depr_args_and_modifier[-1]):
+                    deprecated_args = depr_args_and_modifier[:-1]
+                    _func = depr_args_and_modifier[-1]
+                else:
+                    deprecated_args = depr_args_and_modifier
+
+                    def _func(x): return x
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: this should be a module level (private) function.  It does not need to be re-evaluated every time the decorator is run with a non-callable.

File: coalib/bearlib/__init__.py [hunk 5]
```diff
@@ -82,34 +100,56 @@ def _deprecate_decorator(func):
...
+                        if deprecated_arg not in logged_deprecated_args:
+                            logging.warning(
+                                'The setting `{}` is deprecated. Please use'
+                                ' `{}` instead.'.format(deprecated_arg, arg))
+                            logged_deprecated_args.add(deprecated_arg)
+                        depr_arg_value = modifier.__call__(
+                            kwargs[deprecated_arg])
+                        if arg in kwargs and depr_arg_value != kwargs[arg]:
+                            logging.warning(
+                                'The value of `{}` and `{}` are conflicting.'
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: there is no test coverage of the 'are conflicting' message.  Could you please add it as a doctest in the docstring

File: coalib/bearlib/__init__.py [hunk 6]
```diff
@@ -82,34 +100,56 @@ def _deprecate_decorator(func):
         logged_deprecated_args = set()
 
         def wrapping_function(*args, **kwargs):
-            for arg, depr_arg_and_modifier in depr_args.items():
-                deprecated_arg, _func = (
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: I dont see why you need to modify this line.

`deprecated_arg` will contain a dict, but that can be fixed later.

File: coalib/bearlib/__init__.py [hunk 7]
```diff
@@ -82,34 +100,56 @@ def _deprecate_decorator(func):
...
-                        logged_deprecated_args.add(deprecated_arg)
-                    depr_arg_value = _func.__call__(kwargs[deprecated_arg])
-                    if arg in kwargs and depr_arg_value != kwargs[arg]:
-                        logging.warning(
-                            'The value of `{}` and `{}` are conflicting.'
-                            ' `{}` will be used instead.'.format(
-                                  deprecated_arg, arg, arg))
+            for arg, depr_args_and_modifier in depr_args.items():
+                if not isinstance(depr_args_and_modifier, tuple):
+                    depr_args_and_modifier = (depr_args_and_modifier,)
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: I find it confusing to modify the `key` or `value` variables pulled out of a dict.  When I read `for arg, depr_args_and_modifier in depr_args.items()` at the start of a loop, I expect `depr_args_and_modifier` will always be the same as what occurred in the dict.

File: coalib/bearlib/__init__.py (hunk 6 above)
Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: actually `deprecated_arg` will contain a tuple

File: coalib/bearlib/__init__.py (hunk 5 above)
Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: L57 already covers this

File: coalib/bearlib/__init__.py (hunk 7 above)
Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: Instead of handling the case of one deprecated setting and multiple deprecated settings individually, `depr_args_and_modifier = (depr_args_and_modifier,)` basically just converts the single deprecated setting into a tuple and handles it in the same way.

File: coalib/bearlib/__init__.py (hunk 5 above)
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: :+1:

File: coalib/bearlib/__init__.py (hunk 6 above)
Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: Also each deprecated setting can have its own modifier.

**Comment by pratyushprakash** [comment]
@jayvdb Could you take a look at this?

File: coalib/bearlib/__init__.py (hunk 6 above)
Comment (sils) [inline coalib/bearlib/__init__.py]: @jayvdb is right. You're getting an arg_tuple here just to rename it to depr_args_and_modifier again and never use arg_tuple. That's unneeded complexity.

Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: This was done on request of @jayvdb. He did not want me changing the value pulled out of a dict.

**Comment by pratyushprakash** [comment]
@jayvdb I still don't know what has to be changed. Could you please elaborate your review? Sorry for the trouble but I'm getting conflicting reviews

File: coalib/bearlib/__init__.py (hunk 7 above)
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: For this case, you modify a for loop variable to be a `tuple`, and then unpack that variable, and don't use the for loop variable again.  That is unnecessary complexity.

I believe it can be changed to

```python
    if (isinstance(depr_args_and_modifier, tuple) and
            callable(depr_args_and_modifier[-1])):
        deprecated_args = depr_args_and_modifier[:-1]
        _func = depr_args_and_modifier[-1]
    else:
        deprecated_args = depr_args_and_modifier
        _func = _do_nothing
```

Cleaner, simpler.

**Comment by pratyushprakash** [comment]
I don't know why the circle build always fails.

File: coalib/bearlib/__init__.py [hunk 8]
```diff
@@ -64,12 +67,27 @@ def deprecate_settings(**depr_args):
...
     The metadata for coala has been adjusted as well:
 
     >>> list(run.__metadata__.non_optional_params.keys())
     ['new']
     >>> list(run.__metadata__.optional_params.keys())
-    ['old']
+    ['old', 'older']
 
     You cannot deprecate an already deprecated setting. Don't try. It will
     introduce non-deterministic errors in your program.
```
Comment (sils) [inline coalib/bearlib/__init__.py]: delete this note it's meant for the case you're fixing

Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: @sils Doesn't it still hold. The keyword of the keyword argument to the deprecate_settings decorator must always be the in use setting?

Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: I think the note is warning against the use of `@deprecate_settings(s=(s1,s2),s1=s2)` . Correct me if I am wrong.

Comment (sils) [inline coalib/bearlib/__init__.py]: it should at least be rephrased or removeed IMO

**Comment by pratyushprakash** [comment]
@sils Done

File: coalib/bearlib/__init__.py [hunk 9]
```diff
@@ -82,34 +97,57 @@ def _deprecate_decorator(func):
...
+
             return func(*args, **kwargs)
 
         new_metadata = FunctionMetadata.from_function(func)
-        for arg, depr_arg_and_modifier in depr_args.items():
-            deprecated_arg = (depr_arg_and_modifier[0]
-                              if isinstance(depr_arg_and_modifier, tuple)
-                              else depr_arg_and_modifier)
-            new_metadata.add_deprecated_param(arg, deprecated_arg)
+        for arg, depr_args_and_modifier in depr_args.items():
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: this section of code should be the only code which looks at `depr_args`.
It should save whatever is needed into `__metadata__`, possibly with new fields added.

This chunk of code is run once -- **everything** *possible* should be done here to prepare the decoration, so that each invocation of `wrapping_function` is as fast as possible.

Rename `arg, depr_args_and_modifier` to `arg, depr_value` as its contents & datatype is now not known here

File: coalib/bearlib/__init__.py [hunk 10]
```diff
@@ -82,34 +100,57 @@ def _deprecate_decorator(func):
...
-                            'The setting `{}` is deprecated. Please use `{}` '
-                            'instead.'.format(deprecated_arg, arg))
-                        logged_deprecated_args.add(deprecated_arg)
-                    depr_arg_value = _func.__call__(kwargs[deprecated_arg])
-                    if arg in kwargs and depr_arg_value != kwargs[arg]:
-                        logging.warning(
-                            'The value of `{}` and `{}` are conflicting.'
-                            ' `{}` will be used instead.'.format(
-                                  deprecated_arg, arg, arg))
+            for arg, depr_args_and_modifier in depr_args.items():
```
Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: @jayvdb I don't understand. The previous implementation also iterated through `depr_args`. Also I don't think it's possible to use the iterator provided by `FunctionMetadata` as `wrapping_function.__metadata__.deprecated_params` would only contain the deprecated settings and not the settings that replaced them.

**Comment by JayVB** [comment]
Why am I being added to this thread?!?

**Comment by nemaniarjun** [comment]
@JayVB Someone might have typed your handle instead of @jayvdb

File: coalib/bearlib/__init__.py [hunk 11]
```diff
@@ -64,15 +67,27 @@ def deprecate_settings(**depr_args):
...
+    Hi!
+
     The metadata for coala has been adjusted as well:
 
     >>> list(run.__metadata__.non_optional_params.keys())
     ['new']
     >>> list(run.__metadata__.optional_params.keys())
-    ['old']
-
-    You cannot deprecate an already deprecated setting. Don't try. It will
```
Comment (pratyushprakash) [inline coalib/bearlib/__init__.py]: @sils requested that this be removed.

Comment (jayvdb) [inline coalib/bearlib/__init__.py]: c.f. https://github.com/coala/coala/pull/3546#discussion_r99503527

File: coalib/bearlib/__init__.py [hunk 12]
```diff
@@ -82,34 +97,52 @@ def _deprecate_decorator(func):
...
+                else:
+                    deprecated_arg = depr_value
+                    modifier = (
+                        wrapping_function.__metadata__.default_modifier)
                 if deprecated_arg in kwargs:
                     if deprecated_arg not in logged_deprecated_args:
                         logging.warning(
-                            'The setting `{}` is deprecated. Please use `{}` '
-                            'instead.'.format(deprecated_arg, arg))
+                            'The setting `{}` is deprecated. Please use'
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: why is this modified?

File: coalib/bearlib/__init__.py [hunk 13]
```diff
@@ -82,34 +97,52 @@ def _deprecate_decorator(func):
...
+                            ' `{}` instead.'.format(deprecated_arg, arg))
                         logged_deprecated_args.add(deprecated_arg)
-                    depr_arg_value = _func.__call__(kwargs[deprecated_arg])
+                    depr_arg_value = modifier.__call__(
+                        kwargs[deprecated_arg])
                     if arg in kwargs and depr_arg_value != kwargs[arg]:
                         logging.warning(
                             'The value of `{}` and `{}` are conflicting.'
                             ' `{}` will be used instead.'.format(
-                                  deprecated_arg, arg, arg))
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: this doesnt need changing

File: coalib/bearlib/__init__.py [hunk 14]
```diff
@@ -82,34 +97,52 @@ def _deprecate_decorator(func):
...
+                                deprecated_arg, arg, arg))
                     else:
                         kwargs[arg] = depr_arg_value
                     del kwargs[deprecated_arg]
+
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: while I do prefer a blank line here, I'd much prefer that this patch is shorter so it is easier to see what is being changed.

File: coalib/bearlib/__init__.py [hunk 15]
```diff
@@ -82,34 +97,52 @@ def _deprecate_decorator(func):
...
                 if deprecated_arg in kwargs:
                     if deprecated_arg not in logged_deprecated_args:
                         logging.warning(
-                            'The setting `{}` is deprecated. Please use `{}` '
-                            'instead.'.format(deprecated_arg, arg))
+                            'The setting `{}` is deprecated. Please use'
+                            ' `{}` instead.'.format(deprecated_arg, arg))
                         logged_deprecated_args.add(deprecated_arg)
-                    depr_arg_value = _func.__call__(kwargs[deprecated_arg])
+                    depr_arg_value = modifier.__call__(
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: can you continue using the variable name `_func`, so this change is smaller and more clear.

File: coalib/bearlib/__init__.py [hunk 16]
```diff
@@ -82,34 +97,52 @@ def _deprecate_decorator(func):
         logged_deprecated_args = set()
 
         def wrapping_function(*args, **kwargs):
-            for arg, depr_arg_and_modifier in depr_args.items():
-                deprecated_arg, _func = (
-                    depr_arg_and_modifier
-                    if isinstance(depr_arg_and_modifier, tuple)
-                    else (depr_arg_and_modifier, lambda x: x))
+            for arg, depr_value in wrapping_function.__metadata__.depr_values:
+                if isinstance(depr_value, dict):
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: I think this logic can also be removed into the creation of ` __metadata__` , probably by adding a new optional argument to `add_deprecated_param`

File: coalib/bearlib/__init__.py [hunk 17]
```diff
@@ -82,18 +97,17 @@ def _deprecate_decorator(func):
...
+                deprecated_arg = depr_value[0]
+                _func = depr_value[1]
                 if deprecated_arg in kwargs:
                     if deprecated_arg not in logged_deprecated_args:
                         logging.warning(
                             'The setting `{}` is deprecated. Please use `{}` '
                             'instead.'.format(deprecated_arg, arg))
                         logged_deprecated_args.add(deprecated_arg)
-                    depr_arg_value = _func.__call__(kwargs[deprecated_arg])
+                    depr_arg_value = _func.__call__(
```
Comment (jayvdb) [inline coalib/bearlib/__init__.py]: unnecessary modification

**Comment by pratyushprakash** [comment]
@jayvdb Done

**Comment by pratyushprakash** [comment]
@jayvdb is this ready?

**Comment by jayvdb** [comment]
I do not see a test case for a "A default modifier may be specified at the end of the deprecated settings tuple."

**Comment by pratyushprakash** [comment]
@jayvdb I modified the doctest.

**Comment by jayvdb** [comment]
ack 3cfe7a9

**Comment by jayvdb** [comment]
@rultor merge
